src/backend/distributed/commands/trigger.c

```diff
@@ -203,45 +242,480 @@ get_relation_trigger_oid_compat(HeapTuple heapTuple)
+ * trigger commands.
  */
-void
-ErrorIfUnsupportedCreateTriggerCommand(CreateTrigStmt *createTriggerStmt)
+static void
+ErrorIfUnsupportedTriggerCommand(Node *parseTree, LOCKMODE lockMode)
 {
-	RangeVar *triggerRelation = createTriggerStmt->relation;
+	bool checkTableType = false;
+	RangeVar *relation = NULL;
 
-	bool missingOk = true;
-	Oid relationId = RangeVarGetRelid(triggerRelation, AccessShareLock, missingOk);
+	if (IsA(parseTree, CreateTrigStmt))
+	{
+		/* CREATE TRIGGER triggerName ON relationName */
 
-	if (!OidIsValid(relationId))
+		CreateTrigStmt *createTriggerStmt = castNode(CreateTrigStmt, parseTree);
+		if (!IsCreateCitusTruncateTriggerStmt(createTriggerStmt))
+		{
+			checkTableType = true;
+			relation = createTriggerStmt->relation;
+		}
+	}
+	else if (IsA(parseTree, RenameStmt))
+	{
+		/* ALTER TRIGGER oldName ON relationName RENAME TO newName */
+
+		RenameStmt *renameTriggerStmt = castNode(RenameStmt, parseTree);
+		if (renameTriggerStmt->renameType == OBJECT_TRIGGER)
+		{
+			checkTableType = true;
+			relation = renameTriggerStmt->relation;
+		}
+	}
+	else if (IsA(parseTree, AlterObjectDependsStmt))
+	{
+		/* ALTER TRIGGER triggerName ON relationName DEPENDS ON extensionName */
+
+		AlterObjectDependsStmt *alterTriggerDependsStmt =
+			castNode(AlterObjectDependsStmt, parseTree);
+		if (alterTriggerDependsStmt->objectType == OBJECT_TRIGGER)
+		{
+			checkTableType = true;
+			relation = alterTriggerDependsStmt->relation;
+		}
+	}
+	else if (IsA(parseTree, DropStmt))
+	{
+		/* DROP TRIGGER triggerName ON relationName */
+
+		DropStmt *dropTriggerStmt = castNode(DropStmt, parseTree);
+		if (dropTriggerStmt->removeType == OBJECT_TRIGGER)
+		{
+			checkTableType = true;
+			relation = GetDropTriggerStmtRelation(dropTriggerStmt);
+		}
+	}
+
+	if (!checkTableType)
 	{
-		/*
-		 * standard_ProcessUtility would already error out if the given table
-		 * does not exist
-		 */
 		return;
 	}
 
+	bool missingOk = false;
+	Oid relationId = RangeVarGetRelid(relation, lockMode, missingOk);
+
 	if (!IsCitusTable(relationId))
 	{
 		return;
 	}
 
-	char *functionName = makeRangeVarFromNameList(createTriggerStmt->funcname)->relname;
+	if (!IsCitusLocalTable(relationId))
+	{
+		const char *relationName = relation->relname;
+		ereport(ERROR, (errmsg("cannot execute CREATE/ALTER/DROP trigger command for "
+							   "relation \"%s\" because it is either a distributed "
+							   "table or a reference table", relationName)));
+	}
+
+	EnsureCoordinator();
+}
+
+
+/*
+ * GetDropTriggerStmtRelation takes a DropStmt for a trigger
+ * object and returns RangeVar for the relation that owns the trigger.
+ */
+static RangeVar *
+GetDropTriggerStmtRelation(DropStmt *dropTriggerStmt)
+{
+	Assert(dropTriggerStmt->removeType == OBJECT_TRIGGER);
+
+	List *targetObjectList = dropTriggerStmt->objects;
+
+	/* drop trigger commands can only drop one at a time */
+	Assert(list_length(targetObjectList) == 1);
```

Comment:
Afaik, actually it can't be the case. Postgres already does not allow dropping more than one triggers at a time, not a citus specific behavior

Reply:
Currently it might not be supported by postgres, but postgres doesn't ensure that this behavior won't change in the future, so if we use an error here, we will catch it better if/when that happens. That makes the code more robust IMHO. On the other hand, what is the downside of using an error here?

Reply:
Actually, there is not much downside in erroring out for such cases except the code coverage as we will never hit to this error condition.

However,
1- I cannot see any benefits in erroring out for the cases that are already enforced by postgres and it seems that we prefer assertions instead of errors for such cases.
2- For the ones reading the code, it would be hard to reason why we error out there, is it the citus that doesn't support dropping multiple triggers or is it the postgres ? But development time assertions would already imply that "it is guaranteed but lets catch it in development time, just in case"

Also, the reason that postgres does not support dropping multiple triggers is that triggers do not have unique names, so they should be identified by the owner table.
I think, if they ever bring this uniqueness requirement, it would already be a breaking change

Reply:
Should we remove `bug` keyword from here as well?

Reply:
I think yes